### PR TITLE
Ajd chapter15

### PR DESCRIPTION
**Pull request by alexdeas**
ajd chapter 15

### CODE REVIEW

File: ajd/Animal/lib/Animal.pm [hunk 1]
```diff
@@ -25,6 +25,29 @@ Intended to be an abstract class for other animals
...
+Creates a new animal. Can only be called by sub-classes of Animal.
+
+=cut
+
+sub new {
+
+  my ( $class, $name, $colour ) = @_;
+
+  if ( $class eq 'Animal' ) {
+    die ( "Cannot create an instance of 'Animal', new method can only be called on sub-classes\n" ) ;
```
Comment (dickyince) [inline ajd/Animal/lib/Animal.pm:39]: I forgot this line, doh

File: ajd/Animal/lib/Animal.pm [hunk 2]
```diff
@@ -48,6 +71,70 @@ sub sound {
...
 }
 
+=head2 default_colour
+
+The default colour of an animal. This must be overidden.
+
+=cut
+
+sub default_colour {
+  die 'You must define default_colour() in a subclass';
```
Comment (utopia-2) [inline ajd/Animal/lib/Animal.pm:81]: Didn't the book want the default colour to be brown across _all_ animals, unless overridden?

File: ajd/Animal/lib/Animal.pm [hunk 3]
```diff
@@ -48,6 +71,70 @@ sub sound {
...
+
+=cut
+
+sub name {
+
+  my $self = shift;
+
+  if ( ref( $self ) ) {
+    if ( scalar( @_ ) ) {
+      $self->{name} = shift;
```
Comment (utopia-2) [inline ajd/Animal/lib/Animal.pm:96]: As you're not returning anything anyway, might as well return `$self` for that nice property chaining behaviour:

``` perl
$horsey->name( 'Fancy dancy magic prancy' )->colour( 'Rainbow' );
```

File: ajd/Animal/lib/Animal.pm (hunk 2 above)
Comment (alexdeas) [inline ajd/Animal/lib/Animal.pm:81]: It did that in the chapter, but it wasn't in the exercise.

**Pull request merged**